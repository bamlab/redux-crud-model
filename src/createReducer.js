import R from 'ramda';

export default (initialState, handlers, defaultHandler = state => state) => {
  // initial state is required
  if (R.isNil(initialState)) {
    throw new Error('initial state is required');
  }

  // handlers must be an object
  if (R.isNil(handlers) || !R.is(Object, handlers)) {
    throw new Error('handlers must be an object');
  }

  // handlers cannot have an undefined key
  if (R.any(R.equals('undefined'))(R.keys(handlers))) {
    throw new Error('handlers cannot have an undefined key');
  }

  // create the reducer function
  return (state = initialState, action) => {
    // wrong actions, just return state
    if (R.isNil(action)) return state;
    if (!R.has('type', action)) return state;

    // look for the handler
    const handler = handlers[action.type];

    // no handler use default one
    if (R.isNil(handler)) return defaultHandler(state, action);

    // execute the handler
    return handler(state, action);
  };
};
