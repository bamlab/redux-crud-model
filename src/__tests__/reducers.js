import createReducer from '../reducers';
import createActionTypes from '../actionTypes';

describe('reducers', () => {
  const name = 'myEntity';
  let reducer;
  const actionTypes = createActionTypes({ name });

  beforeEach(() => {
    reducer = createReducer({ name });
  });

  it('should have an initial state', () => {
    const initialState = reducer(undefined, { type: '##INIT' });
    expect(initialState).toEqual({
      entities: {},
      status: {},
    });
  });

  it('should mark entity as fetching', () => {
    const action = { type: actionTypes.fetchOne, payload: { id: 1 } };
    const initialState = reducer(undefined, { type: '##INIT' });
    const state = reducer(initialState, action);
    expect(state).toMatchSnapshot();
  });

  it('should mark entity as failure', () => {
    const action = { type: actionTypes.fetchOneFailure, meta: { id: 1 } };
    const initialState = reducer(undefined, { type: '##INIT' });
    const state = reducer(initialState, action);
    expect(state).toMatchSnapshot();
  });

  it('should handle actions with meta entities key', () => {
    const action = {
      type: 'OtherAction',
      meta: {
        entities: {
          myEntity: {
            'the-id': {
              id: 'the-id',
              foo: 'bar',
            },
          },
        },
      },
    };

    const initialState = reducer(undefined, { type: '##INIT' });
    const state = reducer(initialState, action);
    expect(state).toMatchSnapshot();
  });
});
