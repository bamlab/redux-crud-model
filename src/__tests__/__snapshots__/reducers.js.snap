// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers should handle actions with meta entities key 1`] = `
Object {
  "entities": Object {
    "the-id": Object {
      "foo": "bar",
      "id": "the-id",
    },
  },
  "status": Object {
    "the-id": "fetched",
  },
}
`;

exports[`reducers should mark entity as failure 1`] = `
Object {
  "entities": Object {},
  "status": Object {
    "1": "error",
  },
}
`;

exports[`reducers should mark entity as fetching 1`] = `
Object {
  "entities": Object {},
  "status": Object {
    "1": "fetching",
  },
}
`;
